### PR TITLE
JENKINS-52697: Configuration as Code

### PR DESCRIPTION
**Pull request by mwinter69**
make compatible with configuration as code

### CODE REVIEW

**Comment by jakub-bochenski** [comment]
Please give some better commit message than just "jcasc"

File: src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java [hunk 1]
```diff
@@ -71,10 +72,10 @@ public void validate() throws MimeTypeParseException {
   @DataBoundSetter
   public void setUri(URL url) throws URISyntaxException
   {
-    this.uri = url.toURI();
+    this.uri = url.toExternalForm();
   }
 
-  public void setUri(URI uri) throws URISyntaxException
+  public void setUri(String uri) throws URISyntaxException
```
Comment (jakub-bochenski) [inline src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: Can you add a new setter instead? I don't like changing the field type to String, this should not be neccessary

Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: actually we have to change to string.
Config as code will fail with 
"Cannot find configurator for type class java.net.URI"

Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: I tried locally and it was ok, but next run it failed again. So order is not reliable.

Comment (jakub-bochenski) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: You mean it's not enough to have the String setter defined first?

Not supporting URI fields is strange to me. Is there an open issue about it in JCasc or something?

Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: Obviously not. I know that URL is supported. But I think casc is inspecting the class and finds it is of type URI and then fails.
The change itself is not binary compatible but the configuration is compatible, so no need to do any migration

Comment (jakub-bochenski) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: So what would it take to have JCasc process URI? This seems like a simple change to add support for this type

**Comment by jakub-bochenski** [comment]
> The change itself is not binary compatible but the configuration is compatible, so no need to do any migration

@mwinter69 actually I just realized that while it's XML-configuration compatible, this will still break it for people who use the script configuration approach. I would prefer not to. How about adding a new property-name as a workaround `setUriFromString()` etc?

File: src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java (hunk 1 above)
Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: It is possible to add a converter. CASC uses Stapler which exposes a public static field, where one can register a converter.
With that we can stay with URI

Comment (jakub-bochenski) [reply src/main/java/jenkins/plugins/logstash/configuration/ElasticSearch.java]: Nice, I didn't know about it.

File: src/test/resources/jcasc/elasticSearch.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+jenkins:
```
Comment (darxriggs) [inline src/test/resources/jcasc/elasticSearch.yaml]: Only having everything under `unclassified` would be enough for testing, `jenkins` is not mandatory.

File: src/test/java/jenkins/plugins/logstash/ConfigAsCodeTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import org.jvnet.hudson.test.JenkinsRule;
+
+import io.jenkins.plugins.casc.ConfigurationAsCode;
+import io.jenkins.plugins.casc.ConfiguratorException;
+import jenkins.plugins.logstash.configuration.ElasticSearch;
+import jenkins.plugins.logstash.configuration.Logstash;
+import jenkins.plugins.logstash.configuration.RabbitMq;
+import jenkins.plugins.logstash.configuration.Redis;
+
+public class ConfigAsCodeTest
```
Comment (darxriggs) [inline src/test/java/jenkins/plugins/logstash/ConfigAsCodeTest.java]: I suggest to also implement tests for exporting to YAML.

**Pull request merged**